Breaking change of modal getData()

With the newest version, the return type of `getData()` changed from `any` to `unknown`. This should be included in the list of breaking changes. Why was this change necessary in the first place?

See: [code](https://github.com/maximelafarie/ngx-smart-modal/pull/376/files#diff-778ac4eee25a64f75658af2abbb3812bc68cdbf307fe4501c60c2feeff51a9d2R257)

Hi @maximelafarie, thank you so much for all your amazing work. With v14.0.3, I still see that getData(): unknown and this is causing errors in my build `error TS2339: Property 'data' does not exist on type 'unknown'.`. Are there any plans to add a type instead of unknown?

https://github.com/maximelafarie/ngx-smart-modal/blob/master/projects/ngx-smart-modal/src/lib/components/ngx-smart-modal.component.ts#L261

--- a/projects/ngx-smart-modal/package.json
+++ b/projects/ngx-smart-modal/package.json
@@ -1,6 +1,6 @@
 {
   "name": "ngx-smart-modal",
-  "version": "14.0.3",
+  "version": "14.1.0",
   "description": "Smart modal handler to manage modals and data everywhere in Angular apps.",
   "author": "Maxime LAFARIE <maxime.lafarie@gmail.com>",
   "repository": {

--- a/projects/ngx-smart-modal/src/lib/components/ngx-smart-modal.component.ts
+++ b/projects/ngx-smart-modal/src/lib/components/ngx-smart-modal.component.ts
@@ -25,45 +25,82 @@ import { NgxSmartModalConfig } from '../config/ngx-smart-modal.config';
 @Component({
   selector: 'ngx-smart-modal',
   template: `
-  <div *ngIf="overlayVisible"
-       [style.z-index]="visible ? layerPosition-1 : -1"
-       [ngClass]="{'transparent':!backdrop, 'overlay':true, 'nsm-overlay-open':openedClass}"
-       (click)="dismiss($event)" #nsmOverlay>
-    <div [style.z-index]="visible ? layerPosition : -1"
-         [ngClass]="['nsm-dialog', customClass, openedClass ? 'nsm-dialog-open': 'nsm-dialog-close']" #nsmDialog
-         [attr.aria-hidden]="openedClass ? false : true"
-         [attr.aria-label]="ariaLabel"
-         [attr.aria-labelledby]="ariaLabelledBy"
-         [attr.aria-describedby]="ariaDescribedBy">
-      <div class="nsm-content" #nsmContent>
-        <div class="nsm-body">
-          <ng-template #dynamicContent></ng-template>
-          <ng-content></ng-content>
+    <div
+      *ngIf="overlayVisible"
+      [style.z-index]="visible ? layerPosition - 1 : -1"
+      [ngClass]="{
+        transparent: !backdrop,
+        overlay: true,
+        'nsm-overlay-open': openedClass
+      }"
+      (click)="dismiss($event)"
+      #nsmOverlay
+    >
+      <div
+        [style.z-index]="visible ? layerPosition : -1"
+        [ngClass]="[
+          'nsm-dialog',
+          customClass,
+          openedClass ? 'nsm-dialog-open' : 'nsm-dialog-close'
+        ]"
+        #nsmDialog
+        [attr.aria-hidden]="openedClass ? false : true"
+        [attr.aria-label]="ariaLabel"
+        [attr.aria-labelledby]="ariaLabelledBy"
+        [attr.aria-describedby]="ariaDescribedBy"
+      >
+        <div class="nsm-content" #nsmContent>
+          <div class="nsm-body">
+            <ng-template #dynamicContent></ng-template>
+            <ng-content></ng-content>
+          </div>
+          <button
+            type="button"
+            *ngIf="closable"
+            (click)="close()"
+            aria-label="Close"
+            class="nsm-dialog-btn-close"
+          >
+            <svg
+              xmlns="http://www.w3.org/2000/svg"
+              version="1.1"
+              id="Layer_1"
+              x="0px"
+              y="0px"
+              viewBox="0 0 512 512"
+              xml:space="preserve"
+              width="16px"
+              height="16px"
+              role="img"
+              aria-labelledby="closeIconTitle closeIconDesc"
+            >
+              <title id="closeIconTitle">Close Icon</title>
+              <desc id="closeIconDesc">
+                A light-gray close icon used to close the modal
+              </desc>
+              <g>
+                <path
+                  d="M505.943,6.058c-8.077-8.077-21.172-8.077-29.249,0L6.058,476.693c-8.077,8.077-8.077,21.172,0,29.249    C10.096,509.982,15.39,512,20.683,512c5.293,0,10.586-2.019,14.625-6.059L505.943,35.306    C514.019,27.23,514.019,14.135,505.943,6.058z"
+                  fill="currentColor"
+                />
+              </g>
+              <g>
+                <path
+                  d="M505.942,476.694L35.306,6.059c-8.076-8.077-21.172-8.077-29.248,0c-8.077,8.076-8.077,21.171,0,29.248l470.636,470.636    c4.038,4.039,9.332,6.058,14.625,6.058c5.293,0,10.587-2.019,14.624-6.057C514.018,497.866,514.018,484.771,505.942,476.694z"
+                  fill="currentColor"
+                />
+              </g>
+            </svg>
+          </button>
         </div>
-        <button type="button" *ngIf="closable" (click)="close()" aria-label="Close" class="nsm-dialog-btn-close">
-          <svg xmlns="http://www.w3.org/2000/svg" version="1.1" id="Layer_1" x="0px" y="0px" viewBox="0 0 512 512"
-               xml:space="preserve" width="16px" height="16px" role="img" aria-labelledby="closeIconTitle closeIconDesc">
-            <title id="closeIconTitle">Close Icon</title>
-            <desc id="closeIconDesc">A light-gray close icon used to close the modal</desc>
-            <g>
-              <path d="M505.943,6.058c-8.077-8.077-21.172-8.077-29.249,0L6.058,476.693c-8.077,8.077-8.077,21.172,0,29.249    C10.096,509.982,15.39,512,20.683,512c5.293,0,10.586-2.019,14.625-6.059L505.943,35.306    C514.019,27.23,514.019,14.135,505.943,6.058z"
-                    fill="currentColor"/>
-            </g>
-            <g>
-              <path d="M505.942,476.694L35.306,6.059c-8.076-8.077-21.172-8.077-29.248,0c-8.077,8.076-8.077,21.171,0,29.248l470.636,470.636    c4.038,4.039,9.332,6.058,14.625,6.058c5.293,0,10.587-2.019,14.624-6.057C514.018,497.866,514.018,484.771,505.942,476.694z"
-                    fill="currentColor"/>
-            </g>
-          </svg>
-        </button>
       </div>
     </div>
-  </div>
-`,
-  styles: [
-  ]
+  `,
+  styles: [],
 })
-export class NgxSmartModalComponent implements OnInit, OnDestroy, AfterViewChecked {
-
+export class NgxSmartModalComponent
+  implements OnInit, OnDestroy, AfterViewChecked
+{
   @Input() public closable = true;
   @Input() public escapable = true;
   @Input() public dismissable = true;
@@ -80,13 +117,15 @@ export class NgxSmartModalComponent implements OnInit, OnDestroy, AfterViewCheck
   @Input() public ariaDescribedBy: string | null = null;
   @Input() public refocus = true;
 
-  @Output() public visibleChange: EventEmitter<boolean> = new EventEmitter<boolean>();
+  @Output() public visibleChange: EventEmitter<boolean> =
+    new EventEmitter<boolean>();
   @Output() public onClose: EventEmitter<any> = new EventEmitter();
   @Output() public onCloseFinished: EventEmitter<any> = new EventEmitter();
   @Output() public onDismiss: EventEmitter<any> = new EventEmitter();
   @Output() public onDismissFinished: EventEmitter<any> = new EventEmitter();
   @Output() public onAnyCloseEvent: EventEmitter<any> = new EventEmitter();
-  @Output() public onAnyCloseEventFinished: EventEmitter<any> = new EventEmitter();
+  @Output() public onAnyCloseEventFinished: EventEmitter<any> =
+    new EventEmitter();
   @Output() public onOpen: EventEmitter<any> = new EventEmitter();
   @Output() public onOpenFinished: EventEmitter<any> = new EventEmitter();
   @Output() public onEscape: EventEmitter<any> = new EventEmitter();
@@ -106,7 +145,8 @@ export class NgxSmartModalComponent implements OnInit, OnDestroy, AfterViewCheck
   @ViewChildren('nsmContent') private nsmContent!: QueryList<ElementRef>;
   @ViewChildren('nsmDialog') public nsmDialog!: QueryList<ElementRef>;
   @ViewChildren('nsmOverlay') private nsmOverlay!: QueryList<ElementRef>;
-  @ViewChild('dynamicContent', { read: ViewContainerRef }) private dynamicContentContainer!: ViewContainerRef;
+  @ViewChild('dynamicContent', { read: ViewContainerRef })
+  private dynamicContentContainer!: ViewContainerRef;
 
   constructor(
     private _renderer: Renderer2,
@@ -114,12 +154,14 @@ export class NgxSmartModalComponent implements OnInit, OnDestroy, AfterViewCheck
     private _viewContainerRef: ViewContainerRef,
     public readonly elementRef: ElementRef,
     @Inject(DOCUMENT) private _document: Document,
-    @Inject(PLATFORM_ID) private _platformId: object,
-  ) { }
+    @Inject(PLATFORM_ID) private _platformId: object
+  ) {}
 
   public ngOnInit() {
     if (!this.identifier || !this.identifier.length) {
-      throw new Error('identifier field isn’t set. Please set one before calling <ngx-smart-modal> in a template.');
+      throw new Error(
+        'identifier field isn’t set. Please set one before calling <ngx-smart-modal> in a template.'
+      );
     }
 
     this._sendEvent('create');
@@ -170,7 +212,10 @@ export class NgxSmartModalComponent implements OnInit, OnDestroy, AfterViewCheck
    * @returns the modal component
    */
   public dismiss(e: MouseEvent): NgxSmartModalComponent {
-    if (!this.dismissable || !(e?.target as Element)?.classList.contains('overlay')) {
+    if (
+      !this.dismissable ||
+      !(e?.target as Element)?.classList.contains('overlay')
+    ) {
       return this;
     }
 
@@ -258,9 +303,9 @@ export class NgxSmartModalComponent implements OnInit, OnDestroy, AfterViewCheck
   /**
    * Retrieve the data attached to the modal instance
    */
-  public getData(): unknown {
+  public getData<T>(): T {
     this.assignComponentDataToModalData(this._componentRef);
-    return this._data;
+    return this._data as T;
   }
 
   /**
@@ -282,7 +327,10 @@ export class NgxSmartModalComponent implements OnInit, OnDestroy, AfterViewCheck
    * @returns the modal component
    */
   public addBodyClass(): NgxSmartModalComponent {
-    this._renderer.addClass(this._document.body, NgxSmartModalConfig.bodyClassOpen);
+    this._renderer.addClass(
+      this._document.body,
+      NgxSmartModalConfig.bodyClassOpen
+    );
 
     return this;
   }
@@ -293,7 +341,10 @@ export class NgxSmartModalComponent implements OnInit, OnDestroy, AfterViewCheck
    * @returns the modal component
    */
   public removeBodyClass(): NgxSmartModalComponent {
-    this._renderer.removeClass(this._document.body, NgxSmartModalConfig.bodyClassOpen);
+    this._renderer.removeClass(
+      this._document.body,
+      NgxSmartModalConfig.bodyClassOpen
+    );
 
     return this;
   }
@@ -301,7 +352,9 @@ export class NgxSmartModalComponent implements OnInit, OnDestroy, AfterViewCheck
   public markForCheck() {
     try {
       this._changeDetectorRef.detectChanges();
-    } catch (e) { /* empty */ }
+    } catch (e) {
+      /* empty */
+    }
 
     this._changeDetectorRef.markForCheck();
   }
@@ -311,7 +364,13 @@ export class NgxSmartModalComponent implements OnInit, OnDestroy, AfterViewCheck
    */
   @HostListener('window:resize')
   public targetPlacement(): boolean | void {
-    if (!this.isBrowser || !this.nsmDialog.length || !this.nsmContent.length || !this.nsmOverlay.length || !this.target) {
+    if (
+      !this.isBrowser ||
+      !this.nsmDialog.length ||
+      !this.nsmContent.length ||
+      !this.nsmOverlay.length ||
+      !this.target
+    ) {
       return false;
     }
     const targetElement = this._document.querySelector(this.target);
@@ -321,29 +380,64 @@ export class NgxSmartModalComponent implements OnInit, OnDestroy, AfterViewCheck
     }
 
     const targetElementRect = targetElement.getBoundingClientRect();
-    const bodyRect = this.nsmOverlay.first.nativeElement.getBoundingClientRect();
+    const bodyRect =
+      this.nsmOverlay.first.nativeElement.getBoundingClientRect();
 
-    const nsmContentRect = this.nsmContent.first.nativeElement.getBoundingClientRect();
-    const nsmDialogRect = this.nsmDialog.first.nativeElement.getBoundingClientRect();
+    const nsmContentRect =
+      this.nsmContent.first.nativeElement.getBoundingClientRect();
+    const nsmDialogRect =
+      this.nsmDialog.first.nativeElement.getBoundingClientRect();
 
-    const marginLeft = parseInt(getComputedStyle(this.nsmContent.first.nativeElement).marginLeft, 10);
-    const marginTop = parseInt(getComputedStyle(this.nsmContent.first.nativeElement).marginTop, 10);
+    const marginLeft = parseInt(
+      getComputedStyle(this.nsmContent.first.nativeElement).marginLeft,
+      10
+    );
+    const marginTop = parseInt(
+      getComputedStyle(this.nsmContent.first.nativeElement).marginTop,
+      10
+    );
 
-    let offsetTop = targetElementRect.top - nsmDialogRect.top - ((nsmContentRect.height - targetElementRect.height) / 2);
-    let offsetLeft = targetElementRect.left - nsmDialogRect.left - ((nsmContentRect.width - targetElementRect.width) / 2);
+    let offsetTop =
+      targetElementRect.top -
+      nsmDialogRect.top -
+      (nsmContentRect.height - targetElementRect.height) / 2;
+    let offsetLeft =
+      targetElementRect.left -
+      nsmDialogRect.left -
+      (nsmContentRect.width - targetElementRect.width) / 2;
 
-    if (offsetLeft + nsmDialogRect.left + nsmContentRect.width + (marginLeft * 2) > bodyRect.width) {
-      offsetLeft = bodyRect.width - (nsmDialogRect.left + nsmContentRect.width) - (marginLeft * 2);
+    if (
+      offsetLeft + nsmDialogRect.left + nsmContentRect.width + marginLeft * 2 >
+      bodyRect.width
+    ) {
+      offsetLeft =
+        bodyRect.width -
+        (nsmDialogRect.left + nsmContentRect.width) -
+        marginLeft * 2;
     } else if (offsetLeft + nsmDialogRect.left < 0) {
       offsetLeft = -nsmDialogRect.left;
     }
 
-    if (offsetTop + nsmDialogRect.top + nsmContentRect.height + marginTop > bodyRect.height) {
-      offsetTop = bodyRect.height - (nsmDialogRect.top + nsmContentRect.height) - marginTop;
+    if (
+      offsetTop + nsmDialogRect.top + nsmContentRect.height + marginTop >
+      bodyRect.height
+    ) {
+      offsetTop =
+        bodyRect.height -
+        (nsmDialogRect.top + nsmContentRect.height) -
+        marginTop;
     }
 
-    this._renderer.setStyle(this.nsmContent.first.nativeElement, 'top', (offsetTop < 0 ? 0 : offsetTop) + 'px');
-    this._renderer.setStyle(this.nsmContent.first.nativeElement, 'left', offsetLeft + 'px');
+    this._renderer.setStyle(
+      this.nsmContent.first.nativeElement,
+      'top',
+      (offsetTop < 0 ? 0 : offsetTop) + 'px'
+    );
+    this._renderer.setStyle(
+      this.nsmContent.first.nativeElement,
+      'left',
+      offsetLeft + 'px'
+    );
   }
 
   private _sendEvent(name: string, extraData?: unknown): boolean {
@@ -353,10 +447,12 @@ export class NgxSmartModalComponent implements OnInit, OnDestroy, AfterViewCheck
 
     const data = {
       extraData: extraData,
-      instance: { id: this.identifier, modal: this }
+      instance: { id: this.identifier, modal: this },
     };
 
-    const event = new CustomEvent(NgxSmartModalConfig.prefixEvent + name, { detail: data });
+    const event = new CustomEvent(NgxSmartModalConfig.prefixEvent + name, {
+      detail: data,
+    });
 
     return window.dispatchEvent(event);
   }
@@ -373,7 +469,9 @@ export class NgxSmartModalComponent implements OnInit, OnDestroy, AfterViewCheck
    */
   private createDynamicContent(): void {
     this.dynamicContentContainer.clear();
-    this._componentRef = this.dynamicContentContainer.createComponent(this.contentComponent);
+    this._componentRef = this.dynamicContentContainer.createComponent(
+      this.contentComponent
+    );
     this.assignModalDataToComponentData(this._componentRef);
     this.markForCheck();
   }
@@ -381,7 +479,9 @@ export class NgxSmartModalComponent implements OnInit, OnDestroy, AfterViewCheck
   /**
    * Assigns the modal data to the ComponentRef instance properties
    */
-  private assignModalDataToComponentData(componentRef: ComponentRef<Component>): void {
+  private assignModalDataToComponentData(
+    componentRef: ComponentRef<Component>
+  ): void {
     if (componentRef) {
       Object.assign(componentRef.instance, this._data);
     }
@@ -390,10 +490,11 @@ export class NgxSmartModalComponent implements OnInit, OnDestroy, AfterViewCheck
   /**
    * Assigns the ComponentRef instance properties to the modal data object
    */
-  private assignComponentDataToModalData(componentRef: ComponentRef<Component>): void {
+  private assignComponentDataToModalData(
+    componentRef: ComponentRef<Component>
+  ): void {
     if (componentRef) {
       Object.assign(this._data, componentRef.instance);
     }
   }
-
 }

